### PR TITLE
cli: allowing configure to make use of gossip.host

### PR DESCRIPTION
**Pull request by kaplanmaxe**
A few (admittedly unideal) machines for firedancer don't have network interfaces on them with a public IP. Think of a machine here with 1 network interface `eth0` where `eth0` only has a private IP. These machines are NAT'd further up the stack (switch / router layer).

It's still possible for a node like this though to gossip and participate in network consensus by manually setting the gossip host, however the `configure` command currently doesn't check if `gossip.host` is set and will error out. This commit fixes that

### CODE REVIEW

File: src/app/fdctl/config.c
Comment (mmcgee-jump) [inline src/app/fdctl/config.c:793]: We probably still want to check it's an external IPv4 if they have specified, eg, 127.0.0.1 in the file for "[gossip.host]" that's still not valid.  You can rewrite for better style, but something like:

```suggestion
    uint gossip_ip_addr = iface_ip;
    int   gossip_ip4_addr = 0;
    if( FD_UNLIKELY( strlen( config>gossip.host )<=15UL ) ) gossip_ip4_addr = fd_cstr_to_ip4_addr( config>gossip.host, &gossip_ip_addr ); /* Only sets gossip_ip_addr if it's a valid IPv4 address, otherwise assume it's a DNS name and 
    if ( FD_UNLIKELY( !fd_ip4_addr_is_public ( gossip_ip_addr ) && config->is_live_cluster && (gossip_ip4_addr || !strcmp( config->gossip.host, "" ) ) )
```

Comment (mmcgee-jump) [reply src/app/fdctl/config.c:793]: (If you do this, I'd also add to the documentation in `default.toml` that `gossip.host` should be an externally visible IP address for a public cluster.

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]

**Comment by kaplanmaxe** [comment]
Thank you for the feedback above @mmcgee-jump . I agreed with everything you said and made those modifications. Below are some outputs of manual tests I ran.

### No gossip.host set with an interface with a private IP

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml
ERR     11-15 12:01:13.250485 102604 f0   0    src/app/fdctl/config.c(799): Trying to use network interface `wlp110s0` for listening to incoming transactions, but it has IPv4 address x.x.x.x which is part of a private network and will not be routable for other Solana network nodes.
```

### Gossip host set with a private IP (127.0.0.1)

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml
ERR     11-15 12:01:02.938692 102461 f0   0    src/app/fdctl/config.c(804): Trying to use gossip host `127.0.0.1` for listening to incoming transactions, which is not a publicly routable IP address, and will not be routable for other Solana network nodes.
```

### Public IP

I set it to `1.1.1.1` and it worked

File: src/app/fdctl/config.c
Comment (kaplanmaxe) [reply src/app/fdctl/config.c]: Thank you again for the feedback. Addressed both concerns in the latest commit and commented results below.

File: src/app/fdctl/config/default.toml
Comment (mmcgee-jump) [inline src/app/fdctl/config/default.toml:230]: I'd make this a more prescriptive, and this file is double spaced.

```suggestion
    # 127.0.0.1.  If connecting to a public cluster like mainnet or
    # testnet, the IP address or hostname should be publicly routable so
    # it can be reached by other nodes and you should not use an address
    # on a local subnet like 192.168.0.1. 
```

File: src/app/fdctl/config.c
Comment (mmcgee-jump) [inline src/app/fdctl/config.c:799]:
```suggestion
        FD_LOG_ERR(( "Trying to use network interface `%s` for listening to incoming transactions, "
                     "but it has IPv4 address " FD_IP4_ADDR_FMT " which is part of a private network "
                     "and will not be routable for other Solana network nodes. If you are running "
                     "behind a NAT and this interface is publicly reachable, you can continue by "
                     "manually specifying the IP address to advertise in your configuration under "
                     "[gossip.host]." ));
```

**Comment by mmcgee-jump** [comment]
Thanks for the change! Sorry to be so pedantic about the messages, but we want these operator facing errors to be very useful.

File: src/app/fdctl/config/default.toml
Comment (kaplanmaxe) [reply src/app/fdctl/config/default.toml]: Good point. Done

**Comment by kaplanmaxe** [comment]
Thank you for the feedback @mmcgee-jump . It's addressed again. Please don't apologize either. I totally get it and 100% agree with you on everything you said. In addition, my C is very rusty and I am a first-time contributor, so I genuinely appreciate your comments :)

**Comment by kaplanmaxe** [comment]
Ah crap sorry, please hold off on re-reviewing. I tried that commit thing from the GitHub UI for the first time and it screwed it up. Will fix

**Comment by kaplanmaxe** [comment]
Ok sorry about that. Just fixed it up and ran manually on a machine without an interface that has a public IP. Replaced the IP with `x.x.x.x` manually myself in the paste. It's ready for review now

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml
ERR     11-15 13:06:11.883494 157378 f0   0    src/app/fdctl/config.c(799): Trying to use network interface `wlp110s0` for listening to incoming transactions, but it has IPv4 address x.x.x.x which is part of a private network and will not be routable for other Solana network nodes. If you are running behind a NAT and this interface is publicly reachable, you can continue by manually specifying the IP address to advertise in your configuration under [gossip.host].
```

**Comment by mmcgee-jump** [comment]
Great thanks! Looks good. Can you squash the commits?

**Comment by kaplanmaxe** [comment]
Squashed!

**Comment by kaplanmaxe** [comment]
Rebased to resolve conflicts after another PR got merged

**Comment by kaplanmaxe** [comment]
Rebased to resolve conflicts after another PR got merged

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]
One more :)

**Comment by kaplanmaxe** [comment]
Thanks again @mmcgee-jump for the feedback. I've now made this logically simpler in my opinion. Sorry for a bunch of back and forth on this. Admittedly, I'm not a C expert by any means and really a devops guy by trade, but doing my best here :) . Really appreciate the comments. I also added in a check to make sure it actually is a valid IP. Results are as follows:

### With `gossip.host` as `1.1.1.1`

Gets past the check

### With `gossip.host` as `127.0.0.1`

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml 
ERR     11-16 13:40:57.878528 3700874 f0   0    src/app/fdctl/config.c(799): Trying to use gossip host 127.0.0.1 for listening to incoming transactions, which is not a publicly routable IP address, and will not be routable for other Solana network nodes
```

### With `gossip.host` as an invalid ip `1.1.1`

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml 
ERR     11-16 13:41:45.002286 3701454 f0   0    src/app/fdctl/config.c(797): configuration specifies invalid gossip host IP address `1.1.1`
```

### With no `gossip.host` specified on a machine with only interfaces with private IPs

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml 
ERR     11-16 13:42:10.675319 3701782 f0   0    src/app/fdctl/config.c(805): Trying to use network interface `wlp0s20f3` for listening to incoming transactions, but it has IPv4 address <redacted> which is part of a private network and will not be routable for other Solana network nodes. If you are running behind a NAT and this interface is publicly reachable, you can continue by manually specifying the IP address to advertise in your configuration under [gossip.host].
```

**Comment by mmcgee-jump** [comment]
Almost :) but can the gossip.host can be a DNS name too, eg `validator-a.comany.com` ? In that case this worked before but wouldn't now. The logic here is, I think,

* If [gossip.host] is a valid IPv4 address, check it for being external.
* Otherwise, if [gossip.host] is non-empty allow it (we can't check a DNS name).
* Otherwise, check the default IPv4 address we read from the interface is external.

**Comment by kaplanmaxe** [comment]
Hey @mmcgee-jump thanks! But I actually tested and it can't.

```
../firedancer/build/native/gcc/bin/fdctl configure init all --config config.toml
ERR     11-16 14:21:25.976620 3720714 f0   0    src/app/fdctl/config.c(797): configuration specifies invalid gossip host IP address `validator-a.comany.com`
```

The reason is because i'm now calling `fd_cstr_to_ip4_addr` [here](https://github.com/firedancer-io/firedancer/pull/957/files#diff-4063f4668e77e6c1002f6c54c0ef5d9fd728a2f3474fd2132abc7a770c033466R796)

**Comment by kaplanmaxe** [comment]
Oh I see what you mean. You mean that you want to allow that. I never thought about that but I'll change that up

**Comment by kaplanmaxe** [comment]
Ok that should do it @mmcgee-jump (hopefully) :) . Used your suggestion [here](https://github.com/firedancer-io/firedancer/pull/957#discussion_r1394332473) on `strlen` which is pretty reasonable IMO for right now.

**Comment by kaplanmaxe** [comment]
Hey @mmcgee-jump , I'm really sorry to ping you on this one, but wanted to see what I can do to make sure we get this one in. I know you guys are super busy, and I know this is probably low on the priority list, but willing to do what I can to make sure we get this one in. ~47% of Solana's testnet runs on machines that won't work without this PR. Many of these are Stackpath machines. As you can see [here](https://www.validators.app/data-centers?network=testnet), a large percentage of Solana's testnet runs on Stackpath, which is almost all VMs and don't contain interfaces with public IPs. This is mainly due to the TdS program. Happy to do what's necessary here to get this one in!

**Comment by mmcgee-jump** [comment]
I think there was still a subtle issue, if they are using a hostname < 15 chars that logic is still wrong. The back and forth is getting a bit long here, so I thought I'd amend your commit but couldn't since it's in your own fork. I remade the commit with your author attribution, see https://github.com/firedancer-io/firedancer/pull/999

**Pull request closed without merging**
